ao-tools/ao-dumplog/ao-dumplog.1
.\"
.\" Copyright © 2009 Keith Packard <keithp@keithp.com>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
.\"
.\"
.TH AO-DUMPLOG 1 "ao-dumplog" ""
.SH NAME
ao-dumplog \- Store flight log from TeleMetrum device
.SH SYNOPSIS
.B "ao-dumplog"
[\-T \fItty-device\fP]
[\--tty \fItty-device\fP]
[\-D \fIaltos-device\fP]
[\--device \fIaltos-device\fP]
[\--R\fP]
[\--remote\fP]
.SH OPTIONS
.TP
\-T tty-device | --tty tty-device
This selects which tty device ao-dumplog uses to communicate with
the target device.
.TP
\-D AltOS-device | --device AltOS-device
Search for a connected device. This requires an argument of one of the
following forms:
.IP
TeleMetrum:2
.br
TeleMetrum
.br
2
.IP
Leaving out the product name will cause the tool to select a suitable
product, leaving out the serial number will cause the tool to match
one of the available devices.
.TP
\-R | --remote
This uses the command radio link to download the log from TeleMetrum
through a TeleDongle.
.SH DESCRIPTION
.I ao-dumplog
downloads the flight log from a connected TeleMetrum device and stores
it to the configured flight log directory using a name of the form
.IP
\fIyyyy\fP-\fImm\fP-\fIdd\fP-serialP-\fIsss\fP-flight-\fIfff\fP.eeprom
.PP
\fIyyyy\fP is the current year
.br
\fImm\fP is the current month
.br
\fIdd\fP is the current day
.br
\fIsss\fP is the device serial number
.br
\fIfff\fP is a flight sequence number (to make filenames unique)
.SH USAGE
.I ao-dumplog
connects to the specified target device and dumps the stored flight
log.
.SH AUTHOR
Keith Packard


ChangeLog
commit 00f49c8fb0aa38331360bdb85c32bcebe60dcdc3
Author: Bdale Garbee <bdale@gag.com>
Date:   Tue Mar 30 23:19:07 2010 -0600

    fix typo

commit f03ed0876c67b58624abf0c14bf73444b0322d3c
Author: Bdale Garbee <bdale@gag.com>
Date:   Tue Mar 30 23:18:37 2010 -0600

    update changelogs for Debian build

commit b41e617080fe825f7810ee5eee52ea37f7618ec6
Merge: 28e40cc df7bda1
Author: Bdale Garbee <bdale@gag.com>
Date:   Tue Mar 30 23:15:32 2010 -0600

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos
    
    Conflicts:
    	ChangeLog

commit 28e40ccfcd80ab8764d4aa235257cea4d193a0c1
Author: Bdale Garbee <bdale@gag.com>
Date:   Tue Mar 30 23:14:47 2010 -0600

    update changelogs for Debian build

commit 20d4d410e0fc04fe192e309811eed6c0194fa5a8
Author: Bdale Garbee <bdale@gag.com>
Date:   Tue Mar 30 23:11:40 2010 -0600

    initial harness for documentation

commit 9801ff7de21027935f52ccabaa3ff157e22d21ce
Author: Bdale Garbee <bdale@gag.com>
Date:   Mon Mar 15 12:04:18 2010 -0600

    move gbp.conf into debian/

commit df7bda1f32b0049c3878c325ea0b55999f3980e3
Merge: 23da4f3 a7042fe
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Mar 12 10:38:26 2010 -0800

    Merge remote branch 'origin/master'

commit 23da4f3bcdd1d780c9e1f6b68ad2fb309fcae6ba
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Mar 12 10:37:32 2010 -0800

    Document the ao-dumplog '--remote' flag for radio-link downloads
    
    The --remote (or -R) flag uses TeleDongle to fetch data over the radio
    command link from TeleMetrum. It's been there for a while, but the man
    page failed to mention it.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 93df98898f8cd199ae13158bc4f65e3494c954ad
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Mar 12 10:35:12 2010 -0800

    Round radio calibration value instead of truncating
    
    The radio calibration function in the bring-up code was truncating the
    radio calibration value which caused the resulting frequency to
    always measure above the target frequency, instead of trying to get as
    close as possible. This change will result in a closer frequency
    match, but may sometimes be below the target frequency.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit a7042fe7d065d1da1252be5ad43e3c9856214dc8
Author: Bdale Garbee <bdale@Deborah.(none)>
Date:   Fri Mar 5 19:11:11 2010 -0800

    update changelogs for Debian build

commit eda5e1166a97766aa22561beaa1086ba4e19ee16
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Mar 4 17:33:39 2010 -0700

    update changelogs for Debian build

commit 6e61170d42936c18cd6efba6f4c14af616a30745
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Feb 28 18:24:46 2010 -0800

    Need to duplicate new altitude conversion code in aoview.
    
    Altitude conversion is now done with a smaller table and interpolation
    instead of a giant table.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit e4e17e6d3844ae682a0e7a9469a522359bac77b6
Author: Bdale Garbee <bdale@gag.com>
Date:   Sat Feb 27 17:36:13 2010 -0700

    update changelogs for Debian build

commit a1478f65538fdaac7b58ffbd958a035b74956099
Merge: 901fce5 bbf8c9f
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 27 15:19:33 2010 -0800

    Merge remote branch 'origin/master'

commit 901fce5fe3e2762406079ecaa787612b5ed4b34a
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 27 15:18:47 2010 -0800

    Add .gitignore for ao-bringup
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 7aab73a265841aac817ea34235dd1eb819debf76
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 27 15:14:04 2010 -0800

    Fix up LED colors for each product.
    
    Different products assign different color LEDs to the two available
    LED drivers (P1_0, P1_1). Make the LED color pin assignments
    per-product (in ao_pins.h), then deal with not always having a green LED.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 1d6d8d76ec65373ffee70add75d183f5c4168f61
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 27 15:11:44 2010 -0800

    Eliminate deadlock when writing config from radio link
    
    Writing the 'Saved\r\n' string would fill the packet buffer and cause
    a flush to occur, which would need to wait for the radio link to
    receive and transmit a packet. The radio link always re-fetches the
    radio channel number when lighting up the radio, so it need to look in
    the config space. If the config mutex was held by the config writing
    process while the radio was trying to get the channel number, then
    we'd get a deadlock.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit f1956ebadd7bf46b84a05a0c383b6404ca26b344
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 27 15:10:16 2010 -0800

    Leave .ihx files in the build directory too - easier to debug that way
    
    sdcdb wants the .ihx and .cdb files in the same directory, so humor it
    by copying the .ihx files to the src directory instead of moving them.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit bbf8c9f1748af3f1fac08ddf80ae98da2e9b5727
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Feb 26 15:21:12 2010 -0700

    create a turn on script for lighting up TeleDongle v0.2 boards

commit 461d4a1948e112ec7353caf88967391d876469dd
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Feb 26 10:33:13 2010 -0800

    Add LED test

commit 5d7a3a5fbc0af4621c67a6fd51a9c9d5ae688fa5
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Feb 25 16:40:13 2010 -0800

    Fix and document the ao-rawload --run flag
    
    This allows ram-based programs to be loaded and executed easily.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 1e60deca147c85a064719dfad14ccabd1049bbbd
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Feb 25 16:33:34 2010 -0800

    Allow product names to have suffixes (like board revisions)
    

ao-tools/ao-dumplog/ao-dumplog.c
/*
 * Copyright © 2009 Keith Packard <keithp@keithp.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <string.h>
#include "cc-usb.h"
#include "cc.h"

#define NUM_BLOCK	512

static const struct option options[] = {
	{ .name = "tty", .has_arg = 1, .val = 'T' },
	{ .name = "device", .has_arg = 1, .val = 'D' },
	{ .name = "remote", .has_arg = 1, .val = 'R' },
	{ 0, 0, 0, 0},
};

static void usage(char *program)
{
	fprintf(stderr, "usage: %s [--tty <tty-name>] [--device <device-name>] [-R]\n", program);
	exit(1);
}

static uint8_t
log_checksum(int d[8])
{
	uint8_t	sum = 0x5a;
	int	i;

	for (i = 0; i < 8; i++)
		sum += (uint8_t) d[i];
	return -sum;
}

static const char *state_names[] = {
	"startup",
	"idle",
	"pad",
	"boost",
	"fast",
	"coast",
	"drogue",
	"main",
	"landed",
	"invalid"
};


int
main (int argc, char **argv)
{
	struct cc_usb	*cc;
	char		*tty = NULL;
	char		*device = NULL;
	int		c;
	char		line[8192];
	FILE		*out;
	char		*filename;
	int		serial_number = 0;
	int		flight = 0;
	char		cmd;
	int		tick, a, b;
	int		block;
	int		addr;
	int		received_addr;
	int		data[8];
	int		done;
	int		column;
	int		remote = 0;
	int		any_valid;
	int		invalid;
	char		serial_line[8192];

	while ((c = getopt_long(argc, argv, "T:D:R", options, NULL)) != -1) {
		switch (c) {
		case 'T':
			tty = optarg;
			break;
		case 'D':
			device = optarg;
			break;
		case 'R':
			remote = 1;
			break;
		default:
			usage(argv[0]);
			break;
		}
	}
	if (!tty) {
		if (remote)
			tty = cc_usbdevs_find_by_arg(device, "TeleDongle");
		else
			tty = cc_usbdevs_find_by_arg(device, "TeleMetrum");
	}
	if (!tty)
		tty = getenv("ALTOS_TTY");
	if (!tty)
		tty="/dev/ttyACM0";
	cc = cc_usb_open(tty);
	if (!cc)
		exit(1);
	if (remote)
		cc_usb_open_remote(cc);
	/* send a 'version' command followed by a 'log' command */
	cc_usb_printf(cc, "v\n");
	out = NULL;
	for (;;) {
		cc_usb_getline(cc, line, sizeof (line));
		if (sscanf(line, "serial-number %u", &serial_number) == 1)
			strcpy(serial_line, line);
		if (!strncmp(line, "software-version", 16))
			break;
	}
	if (!serial_number) {
		fprintf(stderr, "no serial number found\n");
		cc_usb_close(cc);
		exit(1);
	}
	printf ("Serial number: %d\n", serial_number);
	done = 0;
	column = 0;
	for (block = 0; !done && block < 511; block++) {
		cc_usb_printf(cc, "e %x\n", block);
		if (column == 64) {
			putchar('\n');
			column = 0;
		}
		putchar('.'); fflush(stdout); column++;
		any_valid = 0;
		for (addr = 0; addr < 0x100;) {
			cc_usb_getline(cc, line, sizeof (line));
			if (sscanf(line, "00%x %x %x %x %x %x %x %x %x",
					  &received_addr,
					  &data[0], &data[1], &data[2], &data[3],
					  &data[4], &data[5], &data[6], &data[7]) == 9)
			{
				if (received_addr != addr)
					fprintf(stderr, "data out of sync at 0x%x\n",
						block * 256 + received_addr);

				if (log_checksum(data) != 0)
					fprintf (stderr, "invalid checksum at 0x%x\n",
						 block * 256 + received_addr);
				else
					any_valid = 1;

				cmd = data[0];
				tick = data[2] + (data[3] << 8);
				a = data[4] + (data[5] << 8);
				b = data[6] + (data[7] << 8);
				if (cmd == 'F') {
					flight = b;
					filename = cc_make_filename(serial_number, flight, "eeprom");
					printf ("Flight:       %d\n", flight);
					printf ("File name:     %s\n", filename);
					out = fopen (filename, "w");
					if (!out) {
						perror(filename);
						exit(1);
					}
					fprintf(out, "%s\n", serial_line);
				}

				if (cmd == 'S' && a <= 8) {
					if (column) putchar('\n');
					printf("%s\n", state_names[a]);
					column = 0;
				}
				if (out) {
					fprintf(out, "%c %4x %4x %4x\n",
						cmd, tick, a, b);
					if (cmd == 'S' && a == 8) {
						fclose(out);
						out = NULL;
						done = 1;
					}
				}
				addr += 8;
			}
		}
		if (!any_valid) {
			fclose(out);
			out = NULL;
			done = 1;
		}
	}
	if (column)
		putchar('\n');
	if (out)
		fclose (out);
	cc_usb_close(cc);
	exit (0);
}
